services/traces/downloader.go

```diff
@@ -0,0 +1,91 @@
+package traces
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/onflow/go-ethereum/common"
+	"github.com/rs/zerolog"
+)
+
+const downloadTimeout = 60 * time.Second
+
+type Downloader interface {
+	// Download traces or returning an error with the failure
+	Download(id common.Hash) (json.RawMessage, error)
+}
+
+var _ Downloader = &GCPDownloader{}
+
+type GCPDownloader struct {
+	client     *storage.Client
+	logger     zerolog.Logger
+	cache      *lru.TwoQueueCache[string, []byte]
+	bucketName string
+}
+
+func NewGCPDownloader(bucketName string, logger zerolog.Logger) (*GCPDownloader, error) {
+	if bucketName == "" {
+		return nil, fmt.Errorf("must provide bucket name")
+	}
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("storage.NewClient: %w", err)
+	}
+
+	// try accessing buckets to validate settings
+	_, err = client.Bucket(bucketName).Attrs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error accessing bucket: %s, make sure bucket exists: %w", bucketName, err)
+	}
+
+	valueCache, err := lru.New2Q[string, []byte](128)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create trace cache: %w", err)
+	}
+
+	return &GCPDownloader{
+		client:     client,
+		logger:     logger,
+		bucketName: bucketName,
+		cache:      valueCache,
+	}, nil
+}
+
+func (g *GCPDownloader) Download(id common.Hash) (json.RawMessage, error) {
+	l := g.logger.With().Str("tx-id", id.String()).Logger()
+	val, ok := g.cache.Get(id.String())
+	if ok {
+		l.Debug().Msg("transaction trace cache hit")
+		return val, nil
+	}
+
+	l.Debug().Msg("downloading transaction trace")
+
+	ctx, cancel := context.WithTimeout(context.Background(), downloadTimeout)
+	defer cancel()
+
+	rc, err := g.client.Bucket(g.bucketName).Object(id.String()).NewReader(ctx)
```

Comment:
does the client validate that the downloaded data matches some hash? It would be good to add a check that allows for validating that the data is correct. that may be built into the client, but not sure.